source/nsdynmemLIB/nsdynmemLIB.c

```diff
@@ -117,146 +114,96 @@ void dev_stat_update(mem_stat_update_t type, int16_t size)
     }
 }
 
-//Fix return value allways good fo aligment
 static int heap_alloc_internal_check(int16_t alloc_size)
 {
-    int temp_int, temp_sector = 0, temp_size;
-    temp_size = sizeof(int);
+    int block_count = 0;
+
     if (heap_main == 0) {
         heap_failure(NS_DYN_MEM_HEAP_SECTOR_UNITIALIZED);
-        alloc_size = 0;
     } else if (alloc_size < 1) {
         heap_failure(NS_DYN_MEM_ALLOCATE_SIZE_NOT_VALID);
-        alloc_size = 0;
-    }
-
-
-    if (alloc_size) {
-        //
-        temp_sector = alloc_size;
-        temp_int = (temp_sector % temp_size);
-        if (temp_int) {
-            temp_sector += (temp_size - temp_int);
-        }
-
-
-        temp_sector /= temp_size;
-
-        temp_int = (heap_size / temp_size);
-        temp_int -= 2;
-        if (temp_sector > temp_int) {
-            heap_failure(NS_DYN_MEM_ALLOCATE_SIZE_NOT_VALID);
-            temp_sector = 0;
-        }
+    } else if (alloc_size > (heap_size - 2 * sizeof(int)) ) {
+        heap_failure(NS_DYN_MEM_ALLOCATE_SIZE_NOT_VALID);
+    } else if (alloc_size) {
+        block_count = (alloc_size + sizeof(int) - 1) / sizeof(int);
     }
-    return temp_sector;
+    return block_count;
 }
 
-static int8_t ns_sector_validate(int *sector_start, int direction)
+// Checks that block length indicators equals
+// Block has format: Size of data area in words | data area in words | Size of data area in words
```

Comment:
Not all of those "in words" can mean the same thing. How about

  size [1 word] | data area [abs(size) words] | size [1 word]